# ĐỒ ÁN THỰC HÀNH
- 19120590 - Huỳnh Thanh Mỹ
- 

# A. Thu nhập dữ liệu
1. Chủ đề của dữ liệu là gì và được lấy từ nguồn nào?
- Tên chủ đề: Mức lương cho công việc Khoa học Dữ liệu (Data Science Job Salaries).
- Nguồn dữ liệu: [tại đây](https://salaries.ai-jobs.net/)
2. Tác giả có cho phép sử dụng dữ liệu này không?
- Toàn bộ bộ dữ liệu được công bố trong phạm vi công cộng theo giấy phép CC0 (Public Domain).
3. Tác giả thu thập dữ liệu như thế nào?
- Trang web cung cấp cho người dùng một biểu mẫu ngắn để trả lời khảo sát về công việc và mức lương hiện tại trong lĩnh vực Khoa học dữ liệu

# B. Khám phá dữ liệu

### Import thư viện

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 1. Đọc dữ liệu

- Tập dữ liệu này nói về Mức lương của các lĩnh vực khoa học dữ liệu khác nhau trong lĩnh vực khoa học dữ liệu. 
- Bộ dữ liệu này tạo điều kiện phân tích chuyên sâu về mối quan hệ giữa tiền lương và các yếu tố khác nhau như mức độ kinh nghiệm, loại việc làm, nơi cư trú của nhân viên, tỷ lệ làm việc từ xa, địa điểm công ty và quy mô công ty. Khám phá tập dữ liệu này có thể hỗ trợ đưa ra các quyết định sáng suốt về lựa chọn nghề nghiệp và vị trí công việc trong lĩnh vực Khoa học dữ liệu.
- Ta sẽ tiến hành đọc dữ liệu vào DataFrame có tên là "df"

In [2]:
df = pd.read_csv("./data/salaries.csv")
df.head()

work_year experience_level employment_type             job_title  salary  \
0       2023               SE              FT          Data Modeler  258700   
1       2023               SE              FT          Data Modeler  146600   
2       2023               MI              FT  Data Science Manager  190000   
3       2023               MI              FT  Data Science Manager  160000   
4       2023               EN              FT      Research Analyst   90000   

  salary_currency  salary_in_usd employee_residence  remote_ratio  \
0             USD         258700                 US             0   
1             USD         146600                 US             0   
2             USD         190000                 US           100   
3             USD         160000                 US           100   
4             USD          90000                 US           100   

  company_location company_size  
0               US            M  
1               US            M  
2               US            M  
3               US            M  
4               US            M

## 2. Kích thước dữ liệu

In [3]:
num_rows, num_cols = df.shape
print("Số dòng:", num_rows)
print("Số cột:", num_cols)

Số dòng: 10456
Số cột: 11


## 3. Ý nghĩa từng dòng dữ liệu
- Mỗi hàng trong tập dữ liệu đại diện cho một hồ sơ cá nhân hoặc một trường hợp cụ thể về mức lương trong lĩnh vực Data Science, với thông tin chi tiết về năm làm việc, kinh nghiệm, loại hình làm việc, chức danh công việc, mức lương, loại tiền tệ, địa chỉ cư trú của nhân viên, tỷ lệ làm việc từ xa, địa điểm công ty, và quy mô của công ty.

## 4. Dữ liệu có các dòng bị lặp không?

In [4]:
have_duplicate_rows = df.duplicated().any()
have_duplicate_rows

True

## 5. Ý nghĩa từng cột dữ liệu

Bộ dữ liệu chứa 11 cột, mỗi cột là:

- work_year: Năm trả lương. 
- experience_level: Mức độ kinh nghiệm trong công việc trong năm 
- employment_type: Loại việc làm cho vai trò 
- job_title: Vai trò đã đảm nhiệm trong năm. 
- salary: Tổng số tiền lương được trả. 
- salary_currency: Đơn vị tiền tệ của tiền lương được trả dưới dạng mã tiền tệ ISO 4217. 
- salary_in_usd: Mức lương tính bằng USD 
- employee_residence: Quốc gia cư trú chính của nhân viên trong năm làm việc dưới dạng mã quốc gia ISO 3166. 
- remote_ratio: Tổng khối lượng công việc được thực hiện từ xa 
- company_location: Quốc gia của văn phòng chính hoặc chi nhánh hợp đồng của người sử dụng lao động 
- company_size: Số người trung bình đã làm việc cho công ty trong năm 

In [5]:
print(f"Dữ liệu có {len(df.columns)} cột")
print("Ý nghĩa các cột:\n", list(df.columns))

Dữ liệu có 11 cột
Ý nghĩa các cột:
 ['work_year', 'experience_level', 'employment_type', 'job_title', 'salary', 'salary_currency', 'salary_in_usd', 'employee_residence', 'remote_ratio', 'company_location', 'company_size']


## 6. Kiểu dữ liệu của mỗi cột dữ liệu. Có cột nào có kiểu dữ liệu không phù hợp không?

In [6]:
df.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

- Các thuộc tính của tập dữ liệu đa phần đã ở định dạng phù hợp.

## 7. Với mỗi cột mang giá trị số, các giá trị được phân bố như thế nào?

In [7]:
numerical_df = df.select_dtypes(include=["number"])
numeric_describe = numerical_df.describe()
print(numeric_describe)

          work_year        salary  salary_in_usd  remote_ratio
count  10456.000000  1.045600e+04   10456.000000   10456.00000
mean    2022.778883  1.704655e+05  149263.499904      36.49101
std        0.506937  4.195989e+05   65269.393738      47.58074
min     2020.000000  1.400000e+04   15000.000000       0.00000
25%     2023.000000  1.050000e+05  104640.250000       0.00000
50%     2023.000000  1.430000e+05  142200.000000       0.00000
75%     2023.000000  1.882750e+05  185900.000000     100.00000
max     2023.000000  3.040000e+07  750000.000000     100.00000


### 7.1. Phần trăm giá trị còn thiếu là bao nhiêu?

In [8]:
numerical_df_null = (numerical_df.isnull().sum() / len(df)) * 100

print("Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính số:")
print(numerical_df_null)

Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính số:
work_year        0.0
salary           0.0
salary_in_usd    0.0
remote_ratio     0.0
dtype: float64


### 7.2. Giá trị nhỏ nhất? Giá trị lớn nhất? Các giá trị này có gì bất thường không?

In [9]:
min_values = numerical_df.min()
max_values = numerical_df.max()
print("Giá trị nhỏ nhất:")
print(min_values)
print("\nGiá trị lớn nhất:")
print(max_values)

Giá trị nhỏ nhất:
work_year         2020
salary           14000
salary_in_usd    15000
remote_ratio         0
dtype: int64

Giá trị lớn nhất:
work_year            2023
salary           30400000
salary_in_usd      750000
remote_ratio          100
dtype: int64


- Các giá trị không có gì là bất thường.

## 8. Với mỗi cột mang giá trị phân loại, các giá trị được phân bố như thế nào?

In [10]:
categorical_df = df.drop(numerical_df, axis=1)
categorical_describe = categorical_df.describe()

print(categorical_describe)

       experience_level employment_type      job_title salary_currency  \
count             10456           10456          10456           10456   
unique                4               4            132              22   
top                  SE              FT  Data Engineer             USD   
freq               7325           10410           2390            9569   

       employee_residence company_location company_size  
count               10456            10456        10456  
unique                 86               75            3  
top                    US               US            M  
freq                 9023             9072         9448  


### 8.1. Phần trăm giá trị còn thiếu là bao nhiêu?

In [11]:
categorical_df_null = (categorical_df.isnull().sum() / len(df)) * 100

print("Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính phân loại:")
print(categorical_df_null)

Tỷ lệ phần trăm giá trị thiếu cho mỗi thuộc tính phân loại:
experience_level      0.0
employment_type       0.0
job_title             0.0
salary_currency       0.0
employee_residence    0.0
company_location      0.0
company_size          0.0
dtype: float64


### 8.2. Mỗi cột có bao nhiêu giá trị khác nhau?

In [12]:
for column in categorical_df.columns:
    unique_values = categorical_df[column].unique()
    print(f"+ Cột {column}: {unique_values}")

+ Cột experience_level: ['SE' 'MI' 'EN' 'EX']
+ Cột employment_type: ['FT' 'PT' 'CT' 'FL']
+ Cột job_title: ['Data Modeler' 'Data Science Manager' 'Research Analyst' 'Data Engineer'
 'Data Analyst' 'ETL Developer' 'Machine Learning Engineer'
 'Data Scientist' 'Data Specialist' 'Data Strategist' 'Prompt Engineer'
 'Business Intelligence Developer' 'Research Engineer'
 'Business Intelligence Analyst' 'Data Science Lead' 'ML Engineer'
 'Analytics Engineer' 'BI Developer' 'Data Quality Manager'
 'Applied Scientist' 'Head of Data' 'Research Scientist' 'Data Architect'
 'Data Science' 'Business Intelligence Engineer' 'Data Science Consultant'
 'Machine Learning Scientist' 'Data Product Manager'
 'Business Intelligence Manager' 'Data Manager' 'Data Analytics Manager'
 'Computer Vision Engineer' 'AI Product Manager' 'Data Analytics Lead'
 'Director of Data Science' 'Data Product Owner' 'Data Developer'
 'Machine Learning Infrastructure Engineer' 'Data Lead' 'AI Engineer'
 'Data Integration Eng

### 8.3. Các giá trị này có gì bất thường không?
- Các giá trị không có gì là bất thường.

# C. Đưa ra câu hỏi có ý nghĩa và trả lời câu hỏi

### Câu hỏi 1: "Tình hình làm việc từ xa (remote work) trong lĩnh vực Data Science phản ánh như thế nào qua các năm, và có sự ảnh hưởng của kích thước công ty không?"

Tiền xử lý:

- Chọn cột liên quan: Chọn cột "work_year", "remote_ratio", và "company_size" từ DataFrame.
- Xử lý giá trị thiếu: Kiểm tra và xử lý giá trị thiếu trong các cột cần thiết.
- Kiểm tra và xử lý dữ liệu không hợp lý: Đảm bảo rằng các giá trị của "remote_ratio" nằm trong phạm vi hợp lý (ví dụ: từ 0 đến 100) và kiểm tra tính hợp lý của "work_year" và "company_size".


Phân tích dữ liệu:

- Phân loại dữ liệu: Phân loại dữ liệu theo năm làm việc và kích thước công ty.
- Tính trung bình remote_ratio: Tính giá trị trung bình của cột "remote_ratio" cho từng năm.
- Tính trung bình remote_ratio theo kích thước công ty: Tính giá trị trung bình của "remote_ratio" cho từng phân khúc kích thước công ty.

In [13]:
# Tiền xử lý
questions1_columns = ["work_year", "remote_ratio", "company_size"]
filtered_df = df[questions1_columns].copy()

# Xử lý giá trị thiếu
filtered_df.dropna(subset=["remote_ratio"], inplace=True)

# Kiểm tra và xử lý giá trị không hợp lý
filtered_df = filtered_df[(filtered_df["remote_ratio"] >= 0) & (filtered_df["remote_ratio"] <= 100)]

# Phân tích dữ liệu
average_remote_by_year = filtered_df.groupby("work_year")["remote_ratio"].mean()
average_remote_by_size = filtered_df.groupby("company_size")["remote_ratio"].mean()

print("Trung bình remote_ratio theo năm:")
print(average_remote_by_year)

print("\nTrung bình remote_ratio theo kích thước công ty:")
print(average_remote_by_size)

Trung bình remote_ratio theo năm:
work_year
2020    62.000000
2021    69.954128
2022    55.208964
2023    31.778665
Name: remote_ratio, dtype: float64

Trung bình remote_ratio theo kích thước công ty:
company_size
L    40.083632
M    35.552498
S    70.760234
Name: remote_ratio, dtype: float64


### Nhận xét 

Theo năm: 
- Từ năm 2020 đến năm 2021, trung bình tỷ lệ làm việc từ xa tăng đáng kể, từ 62% lên 69.95%. Điều này có thể phản ánh xu hướng chung trong ngành Data Science hỗ trợ và khuyến khích làm việc từ xa(do dịch Covid).
- Tuy nhiên, từ năm 2022 đến năm 2023, có một giảm đột ngột xuống còn 31.78%. Điều này có thể là do nhiều yếu tố như thay đổi chính sách công ty, tình hình kinh tế, hoặc yếu tố ngoại vi khác.

Theo kích thước công ty: 
- Các công ty có kích thước nhỏ (S) có trung bình tỷ lệ làm việc từ xa cao nhất, đạt 70.76%. Điều này có thể do các công ty nhỏ linh hoạt hơn trong việc triển khai mô hình làm việc từ xa.
- Các công ty có kích thước lớn (L) và trung bình (M) có tỷ lệ làm việc từ xa tương đối thấp hơn, lần lượt là 40.08% và 35.55%. Có thể do các công ty lớn gặp khó khăn hơn trong việc triển khai và duy trì mô hình làm việc từ xa.

Tổng quan: 
- Cần tìm hiểu, phân tích chi tiết hơn để hiểu rõ hơn về tại sao có sự biến động đột ngột trong năm 2023.
- Có thể đề xuất thêm yếu tố như vùng địa lý, ngành nghề, hoặc các yếu tố ngoại vi khác để có cái nhìn toàn diện hơn về tình hình làm việc từ xa trong lĩnh vực Data Science.

### Câu hỏi 2: "Liệu mức lương trong lĩnh vực Data Science có sự khác biệt đáng kể giữa các quốc gia, và có mối liên quan nào giữa mức lương và kích thước công ty không?"

Tiền xử lý:

- Chọn cột liên quan: Chọn cột "salary_in_usd", "employee_residence", và "company_size" từ DataFrame.
- Xử lý giá trị thiếu: Kiểm tra và xử lý giá trị thiếu trong các cột cần thiết.
- Kiểm tra và xử lý dữ liệu không hợp lý: Đảm bảo rằng "salary_in_usd" không chứa các giá trị âm hoặc không hợp lý. Kiểm tra tính hợp lý của "employee_residence" và "company_size".
  
Phân tích dữ liệu:

- Phân loại dữ liệu: Phân loại dữ liệu theo quốc gia cư trú và kích thước công ty.
- Tính trung bình mức lương theo quốc gia: Tính giá trị trung bình của cột "salary_in_usd" cho từng quốc gia.
- Tính trung bình mức lương theo kích thước công ty: Tính giá trị trung bình của "salary_in_usd" cho từng phân khúc kích thước công ty.

In [14]:
# Tiền xử lý
questions2_columns = ["salary_in_usd", "employee_residence", "company_size"]
salary_df = df[questions2_columns].copy()

# Xử lý giá trị thiếu
salary_df.dropna(subset=["salary_in_usd"], inplace=True)

# Kiểm tra và xử lý giá trị không hợp lý
salary_df = salary_df[salary_df["salary_in_usd"] > 0]

# Phân tích dữ liệu
average_salary_by_country = salary_df.groupby("employee_residence")["salary_in_usd"].mean()
average_salary_by_size = salary_df.groupby("company_size")["salary_in_usd"].mean()

# In kết quả
print("Trung bình mức lương theo quốc gia:")
print(average_salary_by_country)

print("\nTrung bình mức lương theo kích thước công ty:")
print(average_salary_by_size)

Trung bình mức lương theo quốc gia:
employee_residence
AD     50745.000000
AE    100000.000000
AM     33500.000000
AR     56444.444444
AS     45555.000000
          ...      
UG     36000.000000
US    157954.720049
UZ     82000.000000
VN     56733.333333
ZA     62935.400000
Name: salary_in_usd, Length: 86, dtype: float64

Trung bình mức lương theo kích thước công ty:
company_size
L    137034.810036
M    151464.452900
S     87513.847953
Name: salary_in_usd, dtype: float64


### Nhận xét
Trung bình mức lương theo quốc gia: 
- Mức lương trung bình có sự biến động lớn giữa các quốc gia. Ví dụ, mức lương trung bình ở Các Tiểu vương quốc Ả Rập Thống nhất (AE) cao nhất với 100,000 USD, trong khi ở Uganda (UG) thấp nhất với 36,000 USD. Sự chênh lệch này có thể phản ánh sự khác biệt về chi phí sinh sống và thu nhập trung bình giữa các quốc gia.
- Có thể cần phân tích chi tiết hơn để hiểu rõ hơn về các yếu tố ảnh hưởng đến mức lương trong từng quốc gia, chẳng hạn như chính sách lương, thị trường lao động, và mức độ phát triển kinh tế.
  
Trung bình mức lương theo kích thước công ty: 
- Có sự chênh lệch đáng kể về mức lương trung bình giữa các kích thước công ty. Các công ty lớn (L) có mức lương trung bình cao nhất với 137,034.81 USD, theo sau là các công ty trung bình (M) với 151,464.45 USD, và cuối cùng là các công ty nhỏ (S) với 87,513.85 USD.
- Có thể giải thích sự chênh lệch này bằng sự khác biệt về khả năng tài chính, chiến lược lương của từng loại công ty, và đặc điểm của người làm việc trong từng loại công ty.
  
Nhận xét tổng quan: 
- Để hiểu rõ hơn về mức lương, có thể thêm yếu tố như trình độ học vấn, kinh nghiệm làm việc, và loại công việc để có cái nhìn chi tiết hơn về yếu tố ảnh hưởng đến mức lương trong lĩnh vực Data Science.